### ISSUE TITLE
Consider embeds as expressions

### ISSUE THREAD

**Issue by FichteFoll** [BenjaminSchaaf/sbnf#5]
Follow-up of #2. I also "used" a variation in #4 in my experimental syntax, but seeing how #2 has not been implemented, it deserves a separate issue.

Reading the embed examples, I don't like how these specified currently. Stating the escape include as well as escape pattern, i.e. the termination of the context, inside an annotation for the match that enters the regex feels very close to the source sublime-syntax system and not intuitive to sbnf.

Instead, an embed should be specified as an independent expression that is anchored to the previous match internally. 

Compare the following:

```
# Current format
regex{string.quoted, include-prototype: false}
    = `'`{punctuation.definition.string.begin, embed: scope:source.regexp, prototype: regex-escape}
    ;

regex-escape = ( ~`\'`{constant.character.escape} )*
               ~`'`{punctuation.definition.string.end}
             ;

# Suggestion
regex{string.quoted, include-prototype: false}
    = `'`{punctuation.definition.string.begin}
      embed!["scope:source.regexp"]{prototype: regex-escape}
      ~`'`{punctuation.definition.string.end}
    ;

regex-escape = ~`\'`{constant.character.escape};
```

I chose the Rust-inspired macro synta here in the form of `name!`. I'm having troubles with coming up with a consistent syntax for parametrizing, however. Should it use `[]` and a native string type as per the discussion in #4? Or should it use `{}`, but here it wouldn't specify the scope to assign to the match?

**Comment by BenjaminSchaaf** [comment]
On the surface I like the idea, and this was how I initially attempted to
implement it, but after trying to do so I found the current implementation
aligns more with how SBNF works. I'll separate this into two parts: 1. Having embed
be a rule/variable and 2. having the prototype escape be defined by what follows
the embed.

#### 1. Embed as a rule

The first thing to note here is that embed can't be like rules and terminals, in
that it can't be used arbitrarily. None of the following can be valid:

```
example = `'` ( embed!["scope:source.regexp"]
              | embed!["scope:source.javascript"] ) ;
example = `'` ( embed!["scope:source.regexp"]
                embed!["scope:source.javascript"] ) ;
example = `'` embed!["scope:source.regexp"]* ;
example = `'` embed!["scope:source.regexp"]? ;
example = `'` ~embed!["scope:source.regexp"] ;
```

As such, embed only makes sense when implemented as a postfix operator. So that
it can only appear after an expression.

It would also be an incredibly unusual operator, as instead of determining how
the next rule/terminal is matches it changes how previous rules/terminals
function:

```
example = `'` ( 'a' 'b' | 'c' )? embed!["scope:source.regexp"] ;
```

In this example the `'` context must embed a syntax, as the `embed` can follow
it, but then it is *also* embedded in the `b` and `c` contexts afterwards. This
behaviour is unexpected, incredibly hard to reason about and also difficult to
implement.

The only way to avoid this situation is to have `embed` be specifically a
postfix operator on a terminal, which just goes right back to the current
implementation :)

#### 2. Having the prototype escape be defined by what follows an embed

This has a much more feasible implementation, but this syntax massively
restricts what can be done with embed:

```
main = 'inc'{embed: scope:source.regexp, prototype: escape} ;
escape = ~'a' ~'b' ;
```

This example compiles to the following:

```yml
- match: 'inc'
  set: 'scope:source.regexp'
  with_prototype:
    - match: 'a'
      set:
        - match: 'b'
          pop: true
```

Such a prototype simply cannot be expressed if only the terminal following an
embed is used to escape it. Doing any of the following conflicts with existing
syntax:

```
main = 'inc'{embed: scope:source.regexp} ( ~'a' ~'b' ) ;
main = 'inc'{embed: scope:source.regexp} escape ;
escape = ~'a' ~'b' ;
```

---

In relation to my first point, I'd also like to add that there's always the
option of using a special syntax to define the embed, eg:

```
example = 'inc'%escape{scope:source.regexp} ;
escape = ~'a' ~'b' ;
```

This has the same semantics but may be easier to read.

As an aside, I had a similar problem to this with `~` early on in development. I
wanted it to be a rule because it felt like the simple solution. Sadly once I
got to the implementation I found it similarly impossible to have it behave as
expected, and thus it got turned into a prefix operator.

**Comment by FichteFoll** [comment]
I struggled a bit with writing down my thoughts or considerations in this post, but I hope you can understand them regardless. Feel free to ask if something is uncertain.

---

Good arguments, but I'm not giving up just yet.

My proposal to solve the first problem is to simply reverse the specification order. Instead of adding an embed action to a rule, the embed macro declares a rule based on a triggering rule. With that in place, it also seem natural to include the escape rule in the same macro call, which could look as follows:

```
regex{string.quoted, include-prototype: false}
  = embed![
      `'`{punctuation.definition.string.begin},
      scope:source.regexp,
      ~`'`{punctuation.definition.string.end}
    ]{prototype: regex-escape}
  ;

regex-escape = ~`\'`{constant.character.escape};
```

Now, the focus here is still on *embedding* something. The sbnf syntax would support the following:
```
embed![`inc`, scope:source.regexp, ~'a' ~'b']
```
But a complex escape rule like this wouldn't be able to be encoded using sublime-syntax's `embed` feature and would thus fail to compile. To achieve a protoype escape equivalent to the given example, the following could be used:

```
embed![`inc`, scope:source.regexp]{prototype: (~'a' ~'b')}
```

Now we have a prototype, but we don't have a proper exit condition for our rule anymore, making this construct rather awkward.

---

It appears that you are instead coming from the opposite direction and treat everything as a prototype push that could be simplified to an embed if the rules are simple enough. The advantage of that is that the decision for which construct to use is abstracted away. The disadvantage is that it may not be clear to SBNF what the user's intent is, because the concept of an embed and a with_prototype push are entirely different.

Let's assume the following example with the example compilation given by you above:
```
main = 'inc'{embed: scope:source.regexp, prototype: escape} ;
escape = ~'a' ~'b' ;
```

It is clear that `escape` is a prototype pattern. However, its exact behavior may be confusing, as the fact that the currently active context is popped of the stack after `escape` has matched *can* clear the embed, but only so if the pushed context has not pushed into another context. Otherwise we would still be somewhere inside the `source.regexp` context but one level higher.

Now, let's assume using `escape = ~'a'`` resulted in the following compiled syntax:
```yaml
- match: 'inc'
  set: 'scope:source.regexp'
  with_prototype:
    - match: 'a'
      pop: true
```
The implementation remains the same as for the above case and it would not guarantee a pop out of the embed. The only way in which such a prototype is able to pop the embed under any condition is using a look-ahead, i.e. `escape = ~'(?=a)'` and this would then be equivalent to using the sublime-text embed action, but the specification seems rather intransparent and hard to decouple from implementation details. Besides, we can't assign scopes within look-aheads (technically we can, but really shouldn't).

**Comment by BenjaminSchaaf** [comment]
> Good arguments, but I'm not giving up just yet.

I'm going to leave my reply/thoughts on the syntax change for another day, but love the enthusiasm :)
It's important to note that the proposed syntax change does not change the semantics of an embed, as it's still essentially sugar for the current syntax.

---

> It is clear that escape is a prototype pattern. However, its exact behavior may be confusing, as the fact that the currently active context is popped of the stack after escape has matched can clear the embed, but only so if the pushed context has not pushed into another context. Otherwise we would still be somewhere inside the source.regexp context but one level higher.

I'm not sure what you're saying here, are you saying that if you `set` the html syntax and it `set`s the JS syntax, that the prototype given with the first `set` will only pop off JS? I was under the impression that `embed`/`escape` always has an equivalent `set`/`with_prototype` form (minus not erroring out when the syntax doesn't exist). If that's not the case SBNF will need two separate syntaxes/parameters for each of those.

**Comment by keith-hall** [comment]
This may help in seeing how its not always possible to emulate an `embed` with `with_prototype`:
https://github.com/trishume/syntect/issues/222#issuecomment-434603252

**Comment by BenjaminSchaaf** [comment]
> This may help in seeing how its not always possible to emulate an `embed` with `with_prototype`:
> [trishume/syntect#222 (comment)](https://github.com/trishume/syntect/issues/222#issuecomment-434603252)

Interesting. I think that solidifies having two separate syntaxes for using `embed` and `with_prototype`.

**Comment by FichteFoll** [comment]
Yes, that is a good example that shows the difference in behavior. If they didn't differ, we wouldn't need `embed` at all, however it has greatly simplified many usages of embedding other syntaxes or contexts and even made some things possible that were not before (those cases outlined in @keith-hall's example). Additionally, it is much easier on resource usage internally, as entire syntax trees don't need to be copied and modified and a syntax can still be parsed correctly when an embedded context cannot be found because the outward behavior does not depend on which context is embedded.

**Comment by BenjaminSchaaf** [comment]
Given the development with #4, here's a suggestion:

```
# with_prototype
main = 'inc'{foo}%[escape]{scope:source.regexp} ;
escape = ~'a' ~'b' ;

# escape
main = 'inc'{foo}%['a+']{scope:source.regexp} ;

# or with keywords

main = 'inc'{foo} % include[escape]{scope:source.regexp} ;
escape = ~'a' ~'b' ;

main = 'inc'{foo} % embed['a+']{scope:source.regexp} ;
```

**Comment by FichteFoll** [comment]
I'm assuming that `%` serves as a binary operator between a rule and an "embed rule".

`'inc'{foo}%[escape]{scope:source.regexp}` is very sigil-heavy and quite noisy. Most of the current punctuation (including #4) is either more or less intuitive or quite integral to making concise definitions (thinking of `~` here), but this proposal really doesn't open up to me.

Between the first two examples, what are the differences? Does it use `escape` for terminals and `with_prototype` for a rule? I thought we agreed on having an explicit syntax difference between the two because they function differently. Or are the comments just misleading?

Regarding the last two, does `include` mean that the 'source.regexp' context is pushed with 'escape' as its prototype? Compared to that, the `embed` example seems self-explanatory, but the syntax is too close to a parametrized rule reference imo.

Are we focusing on enabling `embed`, `with_prototype`, or both?

**Comment by BenjaminSchaaf** [comment]
> Between the first two examples, what are the differences? Does it use escape for terminals and with_prototype for a rule?

Yes. That's why I also included the keyword based alternative, as I think that's much clearer. It's still explicit which is being used, as a rule would always result in `with_prototype` and a regex always in an `escape`. The two examples are equivalent, with the `escape` rule being used as `with_prototype`.

> Are we focusing on enabling embed, with_prototype, or both?

Considering the vast differences, definitely both. I was hoping to keep things simple and just have `embed` just as an optimization of a general include, but clearly that doesn't work.

**Comment by BenjaminSchaaf** [comment]
My proposed syntax is in version `0.3.1`. I'm not really happy with the state it's in, but it does work. Leaving this issue open until we can come up with something better.
